helper/scaleutils/nomad.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	// If the caller has requested more nodes than we have available once
+	// filtered, adjust the value. This shouldn't cause the whole scaling
+	// action to fail, but we should warn.
+	if num > len(nodes) {
+		si.log.Warn("can only identify portion of requested nodes for removal",
+			"requested", num, "available", len(nodes))
+		num = len(nodes)
+	}
+
+	var out []*api.NodeListStub
+
+	// Iterate through our filtered and sorted list of nodes, selecting the
+	// required number of nodes to scale in.
+	for i := 0; i <= num-1; i++ {
+		si.log.Debug("identified Nomad node for removal", "node_id", nodes[i].ID)
+		out = append(out, nodes[i])
+	}
+
+	return out, nil
+}
+
+func (si *ScaleIn) getRemoteIDMap(nodes []*api.NodeListStub, remoteProvider RemoteProvider) ([]NodeIDMap, error) {
+
+	var idFunc nodeIDMapFunc
+
+	// Depending on the remote provider we are targeting, identify the function
+	// we need to use.
+	switch remoteProvider {
+	case RemoteProviderAWS:
+		idFunc = awsNodeIDMap
+	default:
+		return nil, fmt.Errorf("unsupported remote provider: %q", remoteProvider)
+	}
+
+	var (
+		out  []NodeIDMap
+		mErr *multierror.Error
+	)
+
+	for _, node := range nodes {
+
+		// Read the full node object from the API which will contain the full
+		// information required to identify the remote provider ID.
+		nodeInfo, _, err := si.nomad.Nodes().Info(node.ID, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Use the identification function to attempt to pull the remote
+		// provider ID information from the Node info.
+		id, err := idFunc(nodeInfo)
+		if err != nil {
+			mErr = multierror.Append(mErr, err)
+		}
+
+		// Add a nice log message for the operators so they can see the node
+		// that has been identified if they wish.
+		si.log.Debug("identified remote provider ID for node",
+			"node_id", nodeInfo.ID, "remote_id", id)
+
+		out = append(out, NodeIDMap{NomadID: node.ID, RemoteID: id})
+	}
+
+	return out, mErr.ErrorOrNil()
+}
+
+// drainNodes iterates the provides nodeID list and performs a drain on each
+// one.
+func (si *ScaleIn) drainNodes(ctx context.Context, deadline time.Duration, nodes []NodeIDMap) error {
+
+	// Define a WaitGroup. This allows us to trigger each node drain in a go
+	// routine and then wait for them all to complete before exiting.
+	var wg sync.WaitGroup
+
+	// All nodes to be drained form part of a pool of resource and all should
+	// use the same DrainSpec.
+	drainSpec := api.DrainSpec{Deadline: deadline}
+
+	// Define an error to collect errors from each drain routine.
+	var result error
+
+	for _, node := range nodes {
+
+		// Increment our WaitGroup delta.
+		wg.Add(1)
+
+		// Assign our node to a local variable as we are launching a go routine
+		// from within a for loop.
+		n := node
+
+		// Launch a routine to drain the node. Append any error returned to the
+		// error.
+		go func() {
+			if err := si.drainNode(ctx, &wg, n.NomadID, &drainSpec); err != nil {
```

Comment:
You could call `defer wg.Done()` inside the goroutine so you don't have to pass it over to the `drainNode` method.